Add topology metadata support for topics and users

akselh opened the pull request:
This implement support for #199.  

* **Please check if the PR fulfills these requirements**
- [x] The commit messages are descriptive
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] An issue has been created for the pull requests. Some issues might require previous discussion.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
New feature

* **What is the current behavior?** (You can also link to an open issue here)
Not supported.

* **Does this PR introduce a breaking change?** 
No

closes #199

purbon commented:
Thanks a lot for adding this, I am sure is very valuable for many!

purbon commented on src/main/java/com/purbon/kafka/topology/serdes/TopicCustomDeserializer.java:
```diff
@@ -154,11 +141,30 @@ public TopicImpl deserialize(JsonParser parser, DeserializationContext context)
     return topic;
   }
 
+  private Map<String, String> getMap(JsonNode jsonNode) {
```
cosmetic, but may be name _getOptionalMap_ ? what do you think?

akselh replied:
I would then expect Optional<Map<String, String>>  to be the return type.

purbon replied:
it would imply most probably, so please disregard. It works as we have and is clean. Thanks.

purbon commented:
I did a quick review, only left a single cosmetic naming question. In general, I would say, please write a quick doc reference for this and I will merge it asap. Thanks a lot again.

purbon commented:
Probably a stupid question, do we want to store this metadata as part of the state? thinking out loud for when it is used in remote, if other systems want to pick up the information. Does it make sense?

akselh commented:
> Probably a stupid question, do we want to store this metadata as part of the state? thinking out loud for when it is used in remote, if other systems want to pick up the information. Does it make sense?

Have not really thought about that idea. Our use cases are:

1. Validations that includes metadata.
1. Generating documentation from the yaml-files

The way we are thinking about using metadata would not really benefit from having metadata stored in backend state (even though we had some kind of use case). To avoid duplication we are planning to have some of the metadata just as constants in the topology and then have our own separate yaml files with more information related to the metadata.

Maybe an interesting thing could be to provide services for topology retrieval on the KTB-server that you are working?

akselh commented:
A thanks a lot for the quick review! I will add documentation.

But now I see another issue related to validation of metadata: The validation feature only supports validation classes in the Java package "com.purbon.kafka.topology.validation". I think this should be changed to allow specifying of the fully-qualified class name in the config. Then you can add your own custom validations to the classpath when running KTB. I guess for backward compatibility the existing config regime needs to be kept.

Seems right to fix that in this PR as well. What do you think?

purbon commented:
> Seems right to fix that in this PR as well. What do you think?

please do another PR for this.

akselh commented:
> > Seems right to fix that in this PR as well. What do you think?
> 
> please do another PR for this.

Ok.

purbon commented:
Thanks @akselh , I just try to have small things to help review ;-) hope makes sense to you.

akselh commented:
> Thanks @akselh , I just try to have small things to help review ;-) hope makes sense to you.

Yes, actually I totally agree :-) Just had that little thought yesterday...

akselh commented:
@purbon, I created #201 for the issues with validation. Will look into that after I finishing this PR. Would be great to have your input on the issues mentioned in #201 first though! :-)

akselh commented:
@purbon , then I have added documentation. 

Side note: I also added some general docs on what to find in the example directory. When I first started looking into KTB I remember that it wasn't that easy to figure out what can be done in the descriptor files - before I looked into the example directory.

purbon commented:
awesome, thanks a lot @akselh !

purbon reviewed (approved):
LGTM. :shipit:

akselh commented:
Great @purbon , thanks a lot for the fast response and merge! 

I have already started on a code changes for #201, will make PR as soon as I have your feedback on the questions asked :-)
